multicloser/multicloser.go

```diff
@@ -0,0 +1,87 @@
+// Copyright 2023 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package multicloser provides a convenient way to join multiple "close"
+// functions together so they can be called together. This is especially useful
+// to group multiple cleanup function calls and return it as a single "closer"
+// to be called later.
+package multicloser
+
+import (
+	"errors"
+)
+
+// Func is the type signature for a closing function. It accepts a function that
+// returns an error or a void function.
+type Func interface {
+	func() error | func()
+}
+
+// Closer maintains the ordered list of closing functions. Functions will be run
+// in the order in which they were inserted.
+//
+// It is not safe to use concurrently without locking.
+type Closer struct {
+	fns []func() error
+}
+
+// Append adds the given closer functions. It handles void and error signatures.
+// Other signatures should use an anonymous function to match an expected
+// signature.
+func Append[T Func](c *Closer, fns ...T) *Closer {
```

Comment:
This works, and it's typesafe, but you can't _blend types_ in the same call.

```go
// Good
c = multicloser.Append(c, func() {}, func() {})

// Good
c = multicloser.Append(c, func() error {}, func() error {})

// BAD!
c = multicloser.Append(c, func() {}, func() error {})
```

Reply:
One possibility: add function WrapVoid function that takes a `func() void` and returns a no-op `func() error`